benches/bin/mvcctxn.rs

```diff
@@ -0,0 +1,472 @@
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+
+            prewrite(
+                db.clone(),
+                &[Mutation::Put((Key::from_raw(key), value))],
+                key,
+                start_ts,
+            );
+            commit(db.clone(), &[Key::from_raw(key)], start_ts, commit_ts);
+        }
+    }
+    db
+}
+
+#[inline]
+fn get(db: Arc<DB>, key: &Key, statistics: &mut Statistics) -> Option<Vec<u8>> {
+    let snapshot = get_snapshot(db.clone());
+    let start_ts = next_ts();
+    let snapstore = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, true);
+    snapstore.get(key, statistics).unwrap()
+}
+
+fn bench_get(db: Arc<DB>, keys: &[Vec<u8>]) -> f64 {
+    let mut fake_statistics = Statistics::default();
+    let mut rng = rand::thread_rng();
+    do_bench(
+        || {
+            let index = rng.gen_range(0, keys.len());
+            let key = Key::from_raw(&keys[index]);
+
+            get(db.clone(), &key, &mut fake_statistics).unwrap()
+        },
+        500000,
+    )
+}
+
+fn bench_set(db: Arc<DB>, keys: &[Vec<u8>], value_len: usize) -> f64 {
+    let mut rng = rand::thread_rng();
+    do_bench(
+        || {
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+            let value = vec![0u8; value_len];
+
+            let key = &keys[rng.gen_range(0, keys.len())];
+
+            prewrite(
+                db.clone(),
+                &[Mutation::Put((Key::from_raw(key), value))],
+                key,
+                start_ts,
+            );
+            commit(db.clone(), &[Key::from_raw(key)], start_ts, commit_ts)
+        },
+        500000,
+    )
+}
+
+fn bench_delete(db: Arc<DB>, keys: &[Vec<u8>]) -> f64 {
+    let mut rng = rand::thread_rng();
+    do_bench(
+        || {
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+
+            let key = &keys[rng.gen_range(0, keys.len())];
+            prewrite(
+                db.clone(),
+                &[Mutation::Delete(Key::from_raw(key))],
+                key,
+                start_ts,
+            );
+            commit(db.clone(), &[Key::from_raw(key)], start_ts, commit_ts)
+        },
+        500000,
+    )
+}
+
+fn bench_batch_set_impl(db: Arc<DB>, keys: &[Vec<u8>], value_len: usize, batch_size: usize) -> f64 {
+    // Avoid writing duplicated keys in a single transaction
+    let mut indices: Vec<_> = (0..keys.len()).collect();
+    let mut rng = rand::thread_rng();
+
+    let mut keys_to_write: Vec<Key> = Vec::with_capacity(batch_size);
+    let mut mutations: Vec<Mutation> = Vec::with_capacity(batch_size);
+
+    do_bench(
+        || {
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+
+            keys_to_write.clear();
+            mutations.clear();
+
+            // Randomly select non-duplicated keys
+            for i in 0..batch_size {
+                let selected = rng.gen_range(i, keys.len());
```

Comment:
Maybe you can shuffle keys before the bench.

Reply:
here I want to select some distinct keys randomly from `keys` in each iteration...